### ISSUE TITLE
Middleware on first load in static pages

### ISSUE THREAD

**Issue by qm3ster** [nuxt/nuxt.js#2367]
Middleware can depend on query string. Middleware runs once per navigation, wherever. This works wonderfully for first SSR load from a living server.
However, for when generating pages, `query` is always `{}`.
So, when a static page is accessed with a query string (from another site, bookmarks, chat) the resulting page is as if there was no query string, since the middleware is pre-executed only.
Hence, middleware depending on query is broken on static load.
What should we do?

**Comment by Mat-thieu** [comment]
Yeah the middleware doesn't fire on first page visit because it's supposed to fire on the backend (which, ofcourse, doesn't exist when generated)

Any workaround to this?

**Comment by Mat-thieu** [comment]
Commented too quickly, you should use a plugin, I solved it like this;

```javascript
// plugins/token.js
export default function ({ app, store }) {
  app.router.onReady(() => console.log(app.router.currentRoute.query))
}
```

Hope it helps :)

**Comment by qm3ster** [comment]
@Mat-thieu that is indeed one of the workarounds.
Myself, I put the logic into the page component's `created` hook.
But a decision on how this should be addressed should be made, for instance by having no-generate middleware that otherwise conforms to the middleware environment, but runs on the client in case of generate.

**Comment by Mat-thieu** [comment]
Can a contributor tell me if this is expected behaviour?

**Comment by qm3ster** [comment]
It's definitely not a bug, if that's what you mean.
It's _accepted_ behavior.

**Comment by Mat-thieu** [comment]
Fair enough, will use plugins then :)

**Comment by teleyinex** [comment]
This has been soooo helpful that it should be in the official docs.

**Comment by Mat-thieu** [comment]
@SnirShechter The code I posted should also work for a redirect, it waits until the router is completely ready.

Are you using this on the initial load or after some navigation?

**Comment by SnirShechter** [comment]
@Mat-thieu 
I'm using it in the initial load(since after that Vue & Vue-Router are up and running).

Thing is, your code doesn't switch layouts. the router redirects to the correct page with the current layout.

Any help here, anyone?
